README.md
# pdfast


ChangeLog.md
# Changelog for pdfast

## Unreleased changes


test/unit/Spec.hs
import           Test.Hspec
import           Control.Exception              ( evaluate )
import           Test.QuickCheck                ( Testable(property) )
import           ParserSpec.EntrieObject        ( entrieObjParseSpecs )

main :: IO ()
main = hspec entrieObjParseSpecs


test/unit/ParserSpec.hs
{-# LANGUAGE OverloadedStrings #-}

module ParserSpec
  ( objectParseSpecs
  )
where

import           Data.ByteString                ( ByteString )
import           Data.Either                    ( isLeft )
import           Data.PDF.Object

import           Test.Hspec
import           Text.Parsec                    ( parse )

objectParseSpecs :: Spec
objectParseSpecs = do
  describe "Parses valid strings to simple objects" $ do
    parseNullObjects
    parseBooleanObjects
    parseNumericInts
    parseNumericReals
  describe "Parses malformed strings that could be objects" $ do
    malformed "True"
    malformed "fals"
    malformedNumbers

wellformed :: Object -> ByteString -> Spec
wellformed obj s =
  it ("Parsing " ++ show s ++ " to an object") $ case parse objectP "" s of
    Right res -> res `shouldBe` obj

parseNullObjects :: Spec
parseNullObjects = wellformed Null "null"

parseBooleanObjects :: Spec
parseBooleanObjects = do
  wellformed (Boolean True)  "true"
  wellformed (Boolean False) "false"

parseNumericInts :: Spec
parseNumericInts = do
  wellformed (NumericInt 123)   "123"
  wellformed (NumericInt 43445) "43445"
  wellformed (NumericInt 17)    "+17"
  wellformed (NumericInt (-98)) "-98"
  wellformed (NumericInt 0)     "0"

parseNumericReals :: Spec
parseNumericReals = do
  wellformed (NumericReal 34.5)     "34.5"
  wellformed (NumericReal (-3.62))  "-3.62"
  wellformed (NumericReal 123.6)    "+123.6"
  wellformed (NumericReal 4.0)      "4."
  wellformed (NumericReal (-0.002)) "-.002"
  wellformed (NumericReal 0.0)      "0.0"

malformed :: ByteString -> Spec
malformed s =
  it ("Parsing " ++ show s ++ " should fail")
    $          isLeft (parse objectP "" s)
    `shouldBe` True

malformedNumbers :: Spec
malformedNumbers = do
  malformed "+"
  malformed "."
  malformed "1.2.3"
  malformed "1..23"
  malformed "--1"


test/unit/ParserSpec/Object.hs
{-# LANGUAGE OverloadedStrings #-}

module ParserSpec.Object
  ( objectParseSpecs
  )
where

import           Control.Monad.IO.Class         ( liftIO )
import           Data.ByteString                ( ByteString )
import qualified Data.ByteString               as B
import           Data.Either                    ( isLeft )
import           Data.PDF.Object

import           Test.Hspec
import           Text.Parsec                    ( parse )

objectParseSpecs :: Spec
objectParseSpecs = do
  describe "Parses valid strings to simple objects" $ do
    parseNullObjects
    parseBooleanObjects
    parseNumericInts
    parseNumericReals
    parseLiteralStrings
    parseHexadecStrings
    parseNames
  describe "Parses malformed strings that could be objects" $ do
    malformed "True"
    malformed "fals"
    malformedNumbers
    malformedLiteralStrings
    malformedHexadecStrings
    malformedNames

wellformed :: Object NameValue -> ByteString -> Spec
wellformed obj s =
  it ("Parsing " ++ show s ++ " to an object") $ case parse objectP "" s of
    Right res -> res `shouldBe` obj
    Left  err -> liftIO $ print err >> undefined

parseNullObjects :: Spec
parseNullObjects = wellformed Null "null"

parseBooleanObjects :: Spec
parseBooleanObjects = do
  wellformed (Boolean True)  "true"
  wellformed (Boolean False) "false"

parseNumericInts :: Spec
parseNumericInts = do
  wellformed (NumericInt 123)   "123"
  wellformed (NumericInt 43445) "43445"
  wellformed (NumericInt 17)    "+17"
  wellformed (NumericInt (-98)) "-98"
  wellformed (NumericInt 0)     "0"

parseNumericReals :: Spec
parseNumericReals = do
  wellformed (NumericReal 34.5)     "34.5"
  wellformed (NumericReal (-3.62))  "-3.62"
  wellformed (NumericReal 123.6)    "+123.6"
  wellformed (NumericReal 4.0)      "4."
  wellformed (NumericReal (-0.002)) "-.002"
  wellformed (NumericReal 0.0)      "0.0"

parseLiteralStrings :: Spec
parseLiteralStrings = mapM_
  (\(s, res) -> wellformed (LiteralStr res) s)
  [ ("(This is a string)", "This is a string")
  , ( "(Strings may contain newlines\nand such.)"
    , "Strings may contain newlines\nand such."
    )
  , ( "(Strings may contain balanced parentheses ())"
    , "Strings may contain balanced parentheses ()"
    )
  , ( "(special characters (*!&}^% and so on).)"
    , "special characters (*!&}^% and so on)."
    )
  , ("()"                     , "")
  , ("( w h i t e\tspace\n)"  , " w h i t e\tspace\n")
  , ("(\\n\\r\\t\\b\\f\\(\\))", "\n\r\t\b\f()")
  , ("(\\9NIN\\E)"            , "NIN")
  , ("(\\053)"                , "+")
  , ("(\\53)"                 , "+")
  , ("(\\053123)"             , "+123")
  , ("(\\0ZERO)"              , "\0ZERO")
  ]

parseHexadecStrings :: Spec
parseHexadecStrings = mapM_
  (\(s, res) -> wellformed (HexadecStr $ B.unpack res) s)
  [ ("<48656C6C6F20576F726C6421>"      , "Hello World!")
  , ("<230A796F0D4C4F2009537721345C67>", "#\nyo\rLO \tSw!4\\g")
  , ("<>"                              , "")
  , ("<6D497845644c6F576552614e645570506552>", "mIxEdLoWeRaNdUpPeR")
  ]

parseNames :: Spec
parseNames = mapM_
  (\(s, res) -> wellformed (Name $ B.unpack res) s)
  [ ("/Name1"                  , "Name1")
  , ("/ASomewhatLongerName"    , "ASomewhatLongerName")
  , ("/A;Name_With-Various***Characters?", "A;Name_With-Various***Characters?")
  , ("/1.2"                    , "1.2")
  , ("/$$"                     , "$$")
  , ("/@pattern"               , "@pattern")
  , ("/Lime#20Green"           , "Lime Green")
  , ("/paired#28#29parentheses", "paired()parentheses")
  , ("/The_Key_of_F#23_Minor"  , "The_Key_of_F#_Minor")
  , ("/A#42"                   , "AB")
  ]

malformed :: ByteString -> Spec
malformed s =
  it ("Parsing " ++ show s ++ " should fail")
    $          isLeft (parse objectP "" s)
    `shouldBe` True

malformedNumbers :: Spec
malformedNumbers = do
  malformed "+"
  malformed "."
  malformed "1.2.3"
  malformed "1..23"
  malformed "--1"

malformedLiteralStrings :: Spec
malformedLiteralStrings = do
  malformed "(missing ending parentheses"
  malformed "(unbalanced (literal)"
  malformed "(UnbAlAncEd (lItErAL) (* (* )"
  malformed "(this ( dummy -->\\) )"
  malformed "(Dummy Ending \\)"

malformedHexadecStrings :: Spec
malformedHexadecStrings = do
  malformed "<3C4E4F454E44212121"
  malformed "<737061 206365>"
  malformed "<72rrr7272525252>"

malformedNames :: Spec
malformedNames = do
  malformed "/"
  malformed "/f#ail"
  malformed "/bad-ending#"
  malformed "/BAD_END#6"


src/Data/PDF/Object.hs
{-# LANGUAGE LambdaCase #-}

module Data.PDF.Object
  ( Object(..)
  , NameValue
  , objectP
  )
where

import           Data.ByteString                ( ByteString )
import           Data.Char
import           Data.Word                      ( Word8 )
import qualified Data.ByteString               as B
import           Data.Map.Strict                ( Map )
import           Text.Parsec

type NameValue = [Word8]

data Object n = Null
              | Boolean Bool
              | NumericInt Int
              | NumericReal Double
              | LiteralStr ByteString
              | HexadecStr [Word8]
              | Name n
              | Array [Object n]
              | Dictionary (Map n (Object n))
              | NumberTree (Map Int (Object n))
              | Stream
              | Indirect
  deriving (Eq, Show)

type ObjectParser n u m = ParsecT ByteString u m (Object n)

-- | Parser for PDF objects.
objectP :: Monad m => ObjectParser NameValue u m
objectP =
  nullP <|> booleanP <|> nameP <|> numericP <|> hexadecStrP <|> literalP

-- | Parser for null object.
nullP :: Monad m => ObjectParser n u m
nullP = string "null" >> return Null

-- | Parser for boolean objects.
booleanP :: Monad m => ObjectParser n u m
booleanP =
  (string "true" >> return (Boolean True))
    <|> (string "false" >> return (Boolean False))

-- | Parser for numerical objects.
numericP :: Monad m => ObjectParser n u m
numericP = (char '-' >> negate' <$> num) <|> (char '+' >> num) <|> num
 where
  negate' (NumericInt  x) = NumericInt (negate x)
  negate' (NumericReal x) = NumericReal (negate x)

  num = sepBy digits (char '.') >>= \case
    [s@(_ : _)] -> return $ NumericInt $ read s
    [[], []]    -> parserZero <?> "number"
    [s1, s2]    -> return $ NumericReal (expos s1 + fracs s2)
    _           -> parserZero <?> "number"

  digits = many $ oneOf "0123456789"
  expos  = value (10 *)
  fracs s = value (/ 10) (reverse s) / 10
  value f = foldl (\acc x -> f acc + fromIntegral x) 0 . map digitToInt

byte :: Char -> Word8
byte = fromIntegral . ord

readOctal :: String -> Word8
readOctal = fromIntegral . foldl (\x c -> x * 8 + digitToInt c) 0

consA :: Applicative f => a -> f [a] -> f [a]
consA x = (<$>) (x :)

count' :: Monad m => Int -> ParsecT s u m a -> ParsecT s u m [a]
count' 0 _ = return []
count' n p = (p >>= flip consA (count' (n - 1) p)) <|> return []

-- | Parser for literal string objects.
literalP :: Monad m => ObjectParser n u m
literalP = LiteralStr <$> literal'
 where
  literal' = between (char '(') (char ')') next
  literal  = B.cons 40 <$> (B.append <$> literal' <*> (B.cons 41 <$> next))

  next     = escape <|> literal <|> regular <|> return B.empty

  escape   = char '\\' >> (seq <|> octal <|> (anyToken >> next))
  seq      = foldl (<|>) parserZero $ map
    (\(c, b) -> char c >> B.cons b <$> next)
    [ ('n' , 10)
    , ('r' , 13)
    , ('t' , 9)
    , ('b' , 8)
    , ('f' , 12)
    , ('(' , 40)
    , (')' , 41)
    , ('\\', 92)
    ]
  octal   = B.cons . readOctal <$> digits <*> next
  digits  = octDigit >>= flip consA (count' 2 octDigit)
  regular = B.cons . byte <$> satisfy (')' /=) <*> next

-- | Parser for hexadecimal string objects.
hexadecStrP :: Monad m => ObjectParser n u m
hexadecStrP = between (char '<') (char '>') (HexadecStr <$> many hexByte)

-- | Consumes two hex digits and returns it as a byte.
hexByte :: Monad m => ParsecT ByteString u m Word8
hexByte = hexi >>= \n -> fromIntegral . (n * 16 +) <$> hexi
  where hexi = digitToInt <$> hexDigit

-- | Parser for name objects with internal atomic value represented by bytes.
nameP :: Monad m => ObjectParser [Word8] u m
nameP = char '/' >> Name <$> many1 (numsign <|> regular)
 where
  numsign = char '#' >> hexByte
  regular = fromIntegral . fromEnum <$> oneOf ['!' .. '~']



src/Data/PDF/EntrieObject.hs
module Data.PDF.EntrieObject
  ( EntrieObj(..)
  , EntrieNameValue
  , entrieObj
  )
where

import           Data.Bifunctor
import           Data.ByteString                ( ByteString )
import           Data.Char
import           Data.Maybe
import           Data.Word                      ( Word8 )
import qualified Data.ByteString               as B
import           Text.Parsec

-- | EntrieObj is an object that is a valid entrie to a dictionary.
data EntrieObj n = Null
                 | Boolean Bool
                 | NumericInt Int
                 | NumericReal Double
                 | LiteralStr ByteString
                 | HexadecStr [Word8]
                 | Name n
                 | IndirectRef Int Int
                 | Array [EntrieObj n]
                 | Dictionary [(n, EntrieObj n)]
  deriving (Eq, Show)

type EntrieObjParser n u m = ParsecT ByteString u m (EntrieObj n)
type EntrieNameValue = [Word8]

-- -----------------------------------------------------------------------------
-- Parsing
-- -----------------------------------------------------------------------------

-- | Parser for PDF objects.
entrieObj :: Monad m => EntrieObjParser EntrieNameValue u m
entrieObj =
  nullParser
    <|> boolean
    <|> name
    <|> ltgt
    <|> literalStr
    <|> array
    <|> try indirectRef
    <|> numeric

-- | Parser for null object.
nullParser :: Monad m => EntrieObjParser n u m
nullParser = string "null" >> return Null

-- | Parser for boolean objects.
boolean :: Monad m => EntrieObjParser n u m
boolean =
  (string "true" >> return (Boolean True))
    <|> (string "false" >> return (Boolean False))

-- | @numeric@ consumes a decimal number and the result is either an @Int@ or a 
-- @Double@ depending on whether or not the number contains a decimal sign.
--
-- If 'd1' and 'd2' are two string of digits ranging from '0' to '9' with no 
-- trailing or leading zeros then the following inputs will be succesfully 
-- parsed:
--
-- >  "d1"                  -- Evaluates as '(Left Int)'
-- >  "d1.", ".d1", "d1.d2" -- Evaluates as '(Right Double)'
-- >
--
-- In addition to this all inputs may begin with the characters plus '+' or 
-- minus '-'. Unless the numbers begins with a minus character the result will
-- be positive.
--
-- __NOTE__: Not sure if numbers must be followed by spaces. Since I beleive 
-- that the array "[1.2.3 ]" should not be interpreted as "[1.2 .3 ]" I have for 
-- the time being disallowed valid inputs to be followed by a decimal sign.
numeric :: Monad m => EntrieObjParser n u m
numeric = liftNum <$> (plusNumeric <|> minusNumeric <|> numericLiteral)
  where liftNum = either NumericInt NumericReal

plusNumeric :: Monad m => ParsecT ByteString u m (Either Int Double)
plusNumeric = char '+' >> numericLiteral

minusNumeric :: Monad m => ParsecT ByteString u m (Either Int Double)
minusNumeric = char '-' >> bimap negate negate <$> numericLiteral

numericLiteral :: Monad m => ParsecT ByteString u m (Either Int Double)
numericLiteral =
  (   (   intLiteral
      >>= \x -> (Right . (fromIntegral x +) <$> decimalDot) <|> return (Left x)
      )
    <|> Right
    <$> decimalDot1
    )
    <* notFollowedBy (char '.')
 where
  decimalDot1 = char '.' >> (decl <$> many1 digit)
  decimalDot  = char '.' >> (decl <$> many digit)
  decl [] = 0
  decl s  = fromIntegral (read s) / (10 ** fromIntegral (length s))

intLiteral :: (Monad m, Read a, Num a) => ParsecT ByteString u m a
intLiteral = read <$> many1 digit

byte :: Char -> Word8
byte = fromIntegral . ord

readOctal :: String -> Word8
readOctal = fromIntegral . foldl (\x c -> x * 8 + digitToInt c) 0

consA :: Applicative f => a -> f [a] -> f [a]
consA x = (<$>) (x :)

manyLimit :: Monad m => Int -> ParsecT s u m a -> ParsecT s u m [a]
manyLimit 0 _ = return []
manyLimit n p = (p >>= flip consA (manyLimit (n - 1) p)) <|> return []

-- | Parser for literal string objects.
literalStr :: Monad m => EntrieObjParser n u m
literalStr = LiteralStr <$> literal'
 where
  literal' = between (char '(') (char ')') next
  paren    = B.cons 40 <$> (B.append <$> literal' <*> (B.cons 41 <$> next))

  next     = escape <|> paren <|> regular <|> return B.empty

  escape   = char '\\' >> (escseq <|> octal <|> (anyToken >> next))
  escseq   = foldl (<|>) parserZero $ map
    (\(c, b) -> char c >> B.cons b <$> next)
    [ ('n' , 10)
    , ('r' , 13)
    , ('t' , 9)
    , ('b' , 8)
    , ('f' , 12)
    , ('(' , 40)
    , (')' , 41)
    , ('\\', 92)
    ]
  octal   = B.cons . readOctal <$> digits <*> next
  digits  = octDigit >>= flip consA (manyLimit 2 octDigit)
  regular = B.cons . byte <$> satisfy (')' /=) <*> next

-- | Parser for name objects with internal atomic value represented by bytes.
--
-- __NOTE__: It is not clear from the standards but looking at some PDF files it
-- seems as '/' and '(' and '[' can be used as delimeters for names. These 
-- examples is from an actual document:
-- >
-- > <</Type/Pages/Count 2/Kids[ 3 0 R 18 0 R] >>
-- >
--
-- This was a key value pair inside a dictionary.
-- >
-- > /Lang(en-GB) 
-- >
name :: Monad m => EntrieObjParser [Word8] u m
name = Name <$> name'

-- | Turns conforming names into lists of @Word8@. It is not a part of the 
-- @name@ parser since dictionaries uses raw name values as keys.
name' :: Monad m => ParsecT ByteString u m [Word8]
name' = char '/' >> many (numsign <|> regular)
 where
  numsign = char '#' >> hexByte
  regular = byte <$> satisfy (`elem` filter (`notElem` "/([") ['!' .. '~'])

-- | Parser for hexadecimal strings and dictionaries. They have a combined 
-- parser since they are both bounded by lesser than '<' and greater than '>'
-- characters.
ltgt :: Monad m => EntrieObjParser EntrieNameValue u m
ltgt = char '<' >> (hexadecStr <|> dictionary)
 where
  hexadecStr = HexadecStr <$> manyTill hexByte (char '>')

  dictionary = Dictionary <$> between
    (char '<' >> spaces)
    (spaces >> string ">>")
    (many ((,) <$> (name' <* spaces) <*> (spaces *> entrieObj)))

-- | Consumes two hex digits and returns it as a byte.
hexByte :: Monad m => ParsecT ByteString u m Word8
hexByte = hexi >>= \n -> fromIntegral . (n * 16 +) <$> hexi
  where hexi = digitToInt <$> hexDigit

indirectRef :: Monad m => EntrieObjParser n u m
indirectRef =
  IndirectRef <$> (intLiteral <* char ' ') <*> (intLiteral <* string " R")

array :: Monad m => EntrieObjParser EntrieNameValue u m
array = Array <$> between (char '[' >> spaces)
                          (spaces >> char ']')
                          (many (entrieObj <* spaces))
